overlord/snapstate/handlers.go

```diff
@@ -476,6 +481,21 @@ func hasAllContentAttrs(st *state.State, snapName string, requiredContentAttrs [
 func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq map[string][]string, userID int, tm timings.Measurer, flags Flags) error {
 	st := t.State()
 
+	// If transactional, use a single lane for all tasks, so when
+	// one fails the changes for all affected snaps will be
+	// undone. Otherwise, have different lanes per snap so
+	// failures only affect the culprit snap.
+	if flags.Transaction == client.TransactionAllSnaps {
+		lanes := t.Lanes()
+		if len(lanes) != 1 {
+			return fmt.Errorf("internal error: more than one lane (%d) on a transactional action", len(lanes))
+		}
+
+		flags.Lane = lanes[0]
+	} else {
+		flags.Transaction = client.TransactionPerSnap
+	}
```

Comment:
this logic replaces the one below using joinLane?

Reply:
Correct, now that `InstallWithDeviceContext` supports lanes, we can use that rather than manually doing it here.